datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -351,56 +468,108 @@ impl CommonSubexprEliminate {
             schema: orig_schema,
             ..
         } = aggregate;
-        let mut expr_stats = ExprStats::new();
-
         // track transformed information
         let mut transformed = false;
 
-        // rewrite inputs
-        let group_arrays = to_arrays(&group_expr, &mut expr_stats, ExprMask::Normal)?;
-        let aggr_arrays = to_arrays(&aggr_expr, &mut expr_stats, ExprMask::Normal)?;
-
         let name_perserver = NamePreserver::new_for_projection();
         let saved_names = aggr_expr
             .iter()
             .map(|expr| name_perserver.save(expr))
             .collect::<Result<Vec<_>>>()?;
 
-        // rewrite both group exprs and aggr_expr
-        let rewritten = self.rewrite_expr(
-            vec![group_expr, aggr_expr],
-            &[&group_arrays, &aggr_arrays],
-            unwrap_arc(input),
-            &expr_stats,
-            config,
-        )?;
-        transformed |= rewritten.transformed;
-        let (mut new_expr, new_input) = rewritten.data;
-
-        // note the reversed pop order.
-        let new_aggr_expr = pop_expr(&mut new_expr)?;
-        let new_group_expr = pop_expr(&mut new_expr)?;
+        let mut expr_stats = ExprStats::new();
+        // rewrite inputs
+        let (group_found_common, group_arrays) =
+            self.to_arrays(&group_expr, &mut expr_stats, ExprMask::Normal)?;
+        let (aggr_found_common, aggr_arrays) =
+            self.to_arrays(&aggr_expr, &mut expr_stats, ExprMask::Normal)?;
+        let (new_aggr_expr, new_group_expr, new_input) =
+            if group_found_common || aggr_found_common {
+                // rewrite both group exprs and aggr_expr
+                let rewritten = self.rewrite_expr(
+                    vec![group_expr.clone(), aggr_expr.clone()],
```

Comment:
Extra clone, but the reason is the same as https://github.com/apache/datafusion/pull/10473/files#r1648985272

Reply:
Given this only happens when we are actually doing a CSE rewrite (and not on all plans) I think it is fine